service/gcs/core/gcs/gcs.go

```diff
@@ -579,40 +587,77 @@ func (c *gcsCore) WaitContainer(id string) (func() int, error) {
 	return f, nil
 }
 
-// WaitProcess returns a function that can be used to successfully wait for a
-// process exit code. On error the pid was not a valid pid.
-func (c *gcsCore) WaitProcess(pid int) (func() int, error) {
+// WaitProcess returns a channel that can be used to wait for the process exit
+// code and a channel that can be used to signal when the waiter has processed
+// the code fully or decided to stop waiting.
+//
+// The second channel must be signaled in either case to keep the wait counts in
+// sync.
+//
+// On error the pid was not a valid pid and no channels will be returned.
+func (c *gcsCore) WaitProcess(pid int) (chan int, chan bool, error) {
 	c.processCacheMutex.Lock()
 	entry, ok := c.processCache[pid]
 	if !ok {
 		c.processCacheMutex.Unlock()
-		return nil, errors.WithStack(gcserr.NewProcessDoesNotExistError(pid))
+		return nil, nil, errors.WithStack(gcserr.NewProcessDoesNotExistError(pid))
 	}
 	c.processCacheMutex.Unlock()
 
-	// Safe to do without lock because this is a write once variable at creation.
+	// If we are an init process waiter increment our count for this waiter.
 	if entry.isInitProcess {
-		// We only `Add` to the writers lock for the 2nd..Nth waiters because we
-		// added the 1st writer at creation to ensure that the container exited
-		// notification wasn't written until at least one `WaitProcess` was called.
-		entry.syncRoot.Lock()
-		if entry.writersCalled {
-			entry.writersWg.Add(1)
-		} else {
-			entry.writersCalled = true
-		}
-		entry.syncRoot.Unlock()
+		entry.writersSyncRoot.Lock()
```

Comment:
Are these locks necessary?

Reply:
Yes. It is not safe in golang to use a `WaitGroup` where you have someone who has called `Wait()` and others who might call `Add/Done`

Reply:
Ah, interesting. Good to know!